ocaml/xapi/repository.ml

```diff
@@ -769,64 +792,157 @@ let set_pending_guidances ~__context ~host ~guidances =
+      | e ->
+          let host' = Ref.string_of host in
+          let msg = ExnHelper.string_of_exn e in
+          error "applying %s livepatch (%s) on host ref='%s failed: %s"
+            component_str (LivePatch.to_string lp) host' msg ;
+          Some (lp, Api_errors.(Server_error (apply_livepatch_failed, [msg])))
+    )
+    livepatches
+
+let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
+    =
+  (* This function runs on coordinator host *)
+  (* Install RPM updates firstly *)
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.Client.Repository.apply ~rpc ~session_id ~host
+  ) ;
+  (* Apply live patches with best effort *)
+  let failed_livepatches = apply_livepatches' ~__context ~host ~livepatches in
+  let succeeded_livepatches =
+    List.filter
+      (fun (lp, _) ->
+        not (List.exists (fun (lp', _) -> lp = lp') failed_livepatches)
+      )
+      livepatches
+  in
+  (* Update states in cache with best effort as well *)
+  Hashtbl.replace updates_in_cache host
+    (`Assoc
+      [
+        ("updates", `List [])
+      ; ("accumulative_updates", `List [])
+      ; ( "livepatches"
+        , `List
+            (List.map
+               (fun (lp, _) -> `String (LivePatch.to_string lp))
+               failed_livepatches
+            )
+        )
+      ]
+      ) ;
+  (* Evaluate immediate/pending guidances *)
+  let immediate_guidances, pending_guidances =
+    let immediate_guidances' =
+      eval_guidances ~updates_info ~updates:acc_rpm_updates ~kind:Recommended
+        ~livepatches:succeeded_livepatches
+    in
+    let pending_guidances' =
+      List.filter
+        (fun g -> not (List.mem g immediate_guidances'))
+        (eval_guidances ~updates_info ~updates:acc_rpm_updates ~kind:Absolute
+           ~livepatches:[]
+        )
+    in
+    match failed_livepatches with
+    | [] ->
+        (immediate_guidances', pending_guidances')
+    | _ :: _ ->
+        (* There is(are) livepatch failure(s):
+         * the host should not be rebooted, and
+         * an extra pending guidance 'RebootHostOnLivePatchFailure' should be set.
+         *)
+        ( List.filter
+            (fun g -> not (g = Guidance.RebootHost))
+            immediate_guidances'
+        , Guidance.RebootHostOnLivePatchFailure :: pending_guidances'
+        )
+  in
+  List.iter
+    (fun g -> debug "immediate_guidance: %s" (Guidance.to_string g))
+    immediate_guidances ;
+  List.iter
+    (fun g -> debug "pending_guidance: %s" (Guidance.to_string g))
+    pending_guidances ;
+  GuidanceSet.assert_valid_guidances immediate_guidances ;
+  set_pending_guidances ~__context ~host ~guidances:pending_guidances ;
+  let warnings =
+    List.map
+      (fun (lp, e) ->
+        let component_str =
+          Livepatch.string_of_component lp.LivePatch.component
+        in
+        match e with
+        | Api_errors.(Server_error (name, l)) ->
+            name :: component_str :: l
```

Comment:
Another rewrite of these errors as a warning... As above, I'm not sure it is worth doing this for API clients, and it looks confusing in the code.